editor/src/clj/editor/lsp.clj

```diff
@@ -442,11 +449,53 @@
                 (lsp.async/with-auto-evaluation-context evaluation-context
                   (reduce #(sync-resource-state! %1 %2 evaluation-context) state polled-resources)))))
 
+(defn- notify-files-changed! [state change-type->resources]
+  {:pre [(every? #{:created :changed :deleted} (keys change-type->resources))]}
+  (let [watcher->server-ins
+        (util/group-into
+          {} []
+          key val
+          (eduction
+            (mapcat
+              (fn [[server {:keys [in]}]]
+                (eduction
+                  (map #(pair % in))
+                  (:watched-files server))))
+            (:server->server-state state)))
+
+        server-in->changes
+        (util/group-into
+          {} []
+          key val
+          (eduction
+            (mapcat
+              (fn [[{:keys [pattern]} server-ins]]
+                (let [re (Globs/toRegexPattern pattern)]
+                  (eduction
+                    (mapcat
+                      (fn [[change-type resources]]
+                        (eduction
+                          (filter #(.matches (.matcher re (resource/proj-path %))))
+                          (mapcat
+                            (fn [resource]
+                              (eduction
+                                (map #(pair % (pair resource change-type)))
+                                server-ins)))
+                          resources)))
+                    change-type->resources))))
+            watcher->server-ins))]
```

Comment:
Very deep nesting! Using `for` instead would make it a lot easier to follow, but I wanted to keep it performant. Maybe we want to create an educed version of `for`? 🤔